File: esprit_D2F-webapp/src/services/formation/FormationService.ts
import { defaultApi as axios } from "@/utils/helpers/httpClient";
import { config } from "@/config/env";
import type { Formation } from "@/models/formation";
const API_URL = `${config.FORMATION_URL}/formation/formations`;

const FormationService = {
  async createFormation(formationData: Record<string, unknown>): Promise<Formation> {
    const response = await axios.post(API_URL, formationData);
    return response.data;
  },

  async getAllFormations(): Promise<Formation[]> {
    const response = await axios.get(API_URL);
    return response.data;
  },

  async getFormationById(id: number | string): Promise<Formation> {
    const response = await axios.get(`${API_URL}/${id}`);
    return response.data;
  },

  async updateFormation(id: number | string, formationData: Record<string, unknown>): Promise<Formation> {
    const response = await axios.put(`${API_URL}/${id}`, formationData);
    return response.data;
  },

  async deleteFormation(id: number | string): Promise<void> {
    await axios.delete(`${API_URL}/${id}`);
  },
};

export default FormationService;

File: esprit_D2F-formation-build/src/main/java/esprit/pfe/serviceformation/services/FormationService.java
package esprit.pfe.serviceformation.services;

import esprit.pfe.serviceformation.dto.CreateFormationRequest;
import esprit.pfe.serviceformation.dto.FormationResponseDTO;
import esprit.pfe.serviceformation.dto.UpdateFormationRequest;
import esprit.pfe.serviceformation.entities.Formation;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service interface for Formation operations.
 * Provides CRUD and advanced business operations for formations.
 */
public interface FormationService {

    /**
     * Create a new formation
     * @param request formation creation request with validation
     * @return created formation as response DTO
     */
    FormationResponseDTO createFormation(CreateFormationRequest request);

    /**
     * Update an existing formation
     * @param id formation ID
     * @param request update request with partial update support
     * @return updated formation as response DTO
     */
    FormationResponseDTO updateFormation(Long id, UpdateFormationRequest request);

    /**
     * Delete (soft delete) a formation
     * @param id formation ID
     */
    void deleteFormation(Long id);

    /**
     * Get formation by ID
     * @param id formation ID
     * @return formation as response DTO
     */
    FormationResponseDTO getFormationById(Long id);

    /**
     * Get all formations with pagination
     * @param pageable pagination parameters
     * @return paginated formations as response DTOs
     */
    Page<FormationResponseDTO> getAllFormations(Pageable pageable);

    /**
     * Get formation by ID with all relations loaded (for detail view)
     * @param id formation ID
     * @return formation with all relations
     */
    Formation getFormationByIdWithAllRelations(Long id);

    /**
     * Recover a soft-deleted formation (restore it)
     * @param id formation ID
     * @return restored formation as response DTO
     */
    FormationResponseDTO recoverDeletedFormation(Long id);

    /**
     * Clone an existing formation with all its properties
     * @param sourceId source formation ID to clone
     * @param newTitle title for the cloned formation
     * @return cloned formation as response DTO
     */
    FormationResponseDTO cloneFormation(Long sourceId, String newTitle);
}
